pkg/idp/provider/github/manifest/flow.go

```diff
@@ -0,0 +1,188 @@
+		renderer:          newRenderer(),
+		port:              c.Port,
+		readHeaderTimeout: c.ReadHeaderTimeout,
+	}, nil
+}
+
+func CreateGithubApp(c Config) (*githubclient.AppConfig, error) {
+	f, err := newFlow(c)
+	if err != nil {
+		return nil, err
+	}
+	err = f.run()
+	if err != nil {
+		return nil, err
+	}
+	return f.result, err
+}
+
+func (f *Flow) run() error {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	var server *http.Server
+	{
+		mux := http.NewServeMux()
+		// Create a form with the app manifest that is to be submitted to github by the user
+		mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+			// After the manifest is created in github we need to redirect back to complete the flow
+			f.manifest.RedirectURL = fmt.Sprintf("http://%s/retrieve", r.Host)
+
+			jsonManifest, err := json.MarshalIndent(f.manifest, "", "  ")
+			if err != nil {
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+			}
+			template := Template{
+				URL:     f.url.String(),
+				Content: string(jsonManifest),
+			}
+			err = f.renderer.Render("submit.tmpl", w, template)
+			if err != nil {
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+			}
+		})
+		// Retrieves the app code after redirecting back to the local server
+		mux.HandleFunc("/retrieve", func(w http.ResponseWriter, r *http.Request) {
+			if f.state != r.URL.Query().Get("state") {
+				http.Error(w, "state does not match", http.StatusInternalServerError)
+				return
+			}
+			code := r.URL.Query().Get("code")
+			if code == "" {
+				http.Error(w, "code was not found", http.StatusInternalServerError)
+				return
+			}
+			client := githubclient.NewClient(nil)
+			app, resp, err := client.Apps.CompleteAppManifest(ctx, code)
+			if err != nil {
+				http.Error(w, fmt.Sprintf("failed to complete github app manifest: %v", err.Error()), http.StatusInternalServerError)
+			}
+			if resp.StatusCode != http.StatusCreated {
+				http.Error(w, fmt.Sprintf("failed to complete github app manifest. Got response %v", resp), http.StatusInternalServerError)
+			}
+			f.result = app
+			http.Redirect(w, r, "/complete", http.StatusFound)
+		})
+		// The flow is completed, show the user some sort of success and cancel the context
+		mux.HandleFunc("/complete", func(w http.ResponseWriter, r *http.Request) {
+			template := Template{
+				Content: f.manifest.Name,
+			}
+			if err := f.renderer.Render("complete.tmpl", w, template); err != nil {
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+			}
+			cancel()
+		})
+
+		server = &http.Server{
+			Addr:              fmt.Sprintf("localhost:%d", f.port),
+			Handler:           mux,
+			ReadHeaderTimeout: f.readHeaderTimeout,
+		}
+	}
+	err := browser.OpenURL(fmt.Sprintf("http://%s", server.Addr))
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		err = server.ListenAndServe()
+		if errors.Is(err, http.ErrServerClosed) {
+			// All good.
+		} else if err != nil {
+			cancel()
+		}
+	}()
```

Comment:
I wonder whether this here could cause problems, in case the browser opens faster than the server can start.

Reply:
I haven't experienced issues but theoretically it may be possible :thinking:

Reply:
I don't know how to properly wait for the webserver to be started.
We could add some arbitrary timeout, before opening the browser, but that's not nice.

I'm fine with keeping it this way. No reason to fix it if the problem never occurs.